src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs

```diff
@@ -288,20 +302,256 @@ private static ProjectItemGroupElement CreateItemGroup(Project project, string f
+
+                    //In case we found a matching package in requestedVersions, the package will be
+                    //top level. If not, then it is transitive, and include-transitive must be used
+                    //to add the package
+                    if (matchingPackages.Count() != 0 || transitive)
+                    {
+                        
+                        var resolvedVersion = library.Version.ToString();
+                        var packageInfo = new PRPackage { name = library.Name, resolvedVersion = library.Version };
+
+                        if (matchingPackages.Count() != 0)
+                        {
+                            var topLevelPackage = matchingPackages.Single();
+
+                            packageInfo = new PRPackage {
+                                name = packageInfo.name,
+                                resolvedVersion = packageInfo.resolvedVersion,
+                                requestedVersion = topLevelPackage.requestedVersion,
+                                projectFileRequestedVersion = topLevelPackage.projectFileRequestedVersion,
+                                autoReference = topLevelPackage.autoReference
+                            };
+
+                            topLevelPackages.Add(packageInfo);
+                        }
+                        else
+                        {
+                            transitivePackages.Add(packageInfo);
+                        }
+
+                        
+                    }
+
+                }
+
+                resultPackages.Add(
+                    target.TargetFramework.GetShortFolderName(),
+                    Tuple.Create(topLevelPackages.AsEnumerable(), transitivePackages.AsEnumerable())
+                );
+            }
+
+            return resultPackages;
+        }
+
+
         /// <summary>
         /// Returns all package references after evaluating the condition on the item groups.
         /// This method is used when we need package references for a specific target framework.
         /// </summary>
-        /// <param name="project">Project for which the package references have to be obtained.
-        /// The project should have the global property set to have a specific framework</param>
-        /// <param name="libraryDependency">Dependency of the package.</param>
-        /// <returns>List of Items containing the package reference for the package.</returns>
+        /// <param name="project">Project for which the package references have to be obtained.</param>
+        /// <param name="packageId">Name of the package. If empty, returns all package references</param>
+        /// <returns>List of Items containing the package reference for the package.
+        /// If the libraryDependency is null then it returns all package references</returns>
+        private static IEnumerable<ProjectItem> GetPackageReferences(Project project, string packageId) {
```

Comment:
I dont understand the reason of this api here... You have the package id which you read from assets file and get the same packageItem from msbuild project, why?

Reply:
I use that to get the requested version from the project file if possible